spec/12-the-scala-standard-library.md

```diff
@@ -642,6 +642,12 @@ object Predef {
   def classOf[T]: Class[T] = null
    // this is a dummy, classOf is handled by compiler.
 
+  // valueOf -----------------------------------------------------------
+
+  /** Retrieve the single value of a type with a unique inhabitant. */
+  @inline def valueOf[T](implicit vt: ValueOf[T]): T {} = vt.value
```

Comment:
@sjrd I think that's incorrect from multiple points of view:

> This would also have a nice benefit that we could simply write classOf[T] instead of classTag[T].runtimeClass when we have an implicit ClassTag[T] in scope.

- Adding `ClassTag[T]` `to `classOf` would be wrong, as `classOf` requires a `ClassTag` only in some cases, it doesn't require a `ClassTag` when the argument is a constant type.

- Same for `Array[T]`.

I agree that `classTag[T].runtimeClass --> classOf[T]`  is desirable. In fact, I think I have an implementation of this lying around somewhere, which is only blocked by an unrelated scalac issue. Nothing in that requires adding `ClassTag` to `classOf`, and it would be wrong for the reasons mentioned.

This eliminates the argument for

> Since we are defining ValueOf today, we can do it the right way, with the implicit materialization be the only builtin, and valueOf[T] being purely user-space.

as the only question remaining now is whether we want to have `valueOf` consistent with `classOf` or doing its own thing. I prefer consistency.

> I do not understand your argument of "bullet-proof". If I have a v: ValueOf[T] at my disposal, there's nothing wrong in passing it explicitly to valueOf[T](v). The type of v already guarantees that v is the right ValueOf[T] to be passed to valueOf[T] anyway.

What's the use-case for that? The compiler is able to create `ValueOf`s for all valid cases out of thin air, I don't see any benefit where the user can pass his own `ValueOf` instance. Either it is identical to the one supplied by the compiler, or there is a serious issue– as the type can't guarantee that it's actually one of the "allowed" singleton types. That's why the `ValueOf` constructor needed to be private (see comment below). Otherwise a user can do 

  class Horse
  val horse = valueOf(new ValueOf(new Horse)) // or just pass null: Horse

which sabotages the whole idea of `valueOf`.